File: Problems/2.add-two-numbers.py
# title: Add Two Numbers
# timestamp: 2024-12-19 16:03:10
# problemUrl: https://leetcode.com/problems/add-two-numbers/
# memory: 17.9 MB
# runtime: 7 ms

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        head = ListNode((l1.val + l2.val) % 10)
        carry = (l1.val + l2.val) // 10
        
        curr, l1, l2 = head, l1.next, l2.next
        while l1 and l2:
            curr.next = ListNode((l1.val + l2.val + carry) % 10)
            carry = (l1.val + l2.val + carry) // 10
            curr = curr.next
            l1, l2 = l1.next, l2.next

        while l1:
            curr.next = ListNode((l1.val + carry) % 10)
            carry = (l1.val + carry) // 10
            curr = curr.next
            l1 = l1.next
        
        while l2:
            curr.next = ListNode((l2.val + carry) % 10)
            carry = (l2.val + carry) // 10
            curr = curr.next
            l2 = l2.next
        
        if carry:
            curr.next = ListNode(carry)
        
        return head

File: Problems/2.add-two-numbers.go
// title: Add Two Numbers
// timestamp: 2024-04-24 00:32:31
// problemUrl: https://leetcode.com/problems/add-two-numbers/
// memory: 4.3 MB
// runtime: 0 ms

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    var carry int
    for p1, p2 := l1, l2; p1 != nil || p2 != nil; p1, p2 = p1.Next, p2.Next {
        val := carry
        if p1 != nil {
            val += p1.Val
        }
        if p2 != nil {
            val += p2.Val
        }
        p1.Val = val % 10
        carry = val / 10
        if p1.Next == nil && (carry != 0 || p2.Next != nil) {
            p1.Next = &ListNode{}
        }
        if p2.Next == nil && p1.Next != nil {
            p2.Next = &ListNode{}
        }
    }

    return l1
}